### PR TITLE
feat: esp32c3 icm42670 I2C example (panics, help needed)

### PR DESCRIPTION
**Pull request by elpiel**
Before adding the rest of the checks like running the example in CI, etc. I wanted to check if such an example is even needed.
The reason I'm opening this PR is that for whatever reason, using the GPIO pins 10 and 8 of the Esp32-c3 Rust board are giving me an error when I try to run this example.

It cannot instantiate the IMU sensor on the board and throws a bus error:
```
! A panic occured in 'examples/i2c_icm42670.rs', at line 44, column 76
 
PanicInfo {
    payload: Any { .. },
    message: Some(
        called `Result::unwrap()` on an `Err` value: BusError(AckCheckFailed),
    ),
    location: Location {
        file: "examples/i2c_icm42670.rs",
        line: 44,
        col: 76,
    },
    can_unwind: true,
}
```

- I've checked the sensor address
- I've tried multiple I2C clock speeds
- Ran with and without enabling the I2C_EXT0 interrupt priority

### Must

- [x] The code compiles without `errors` or `warnings`.
- [ ] ~All examples work.~ Example does not work
- [ ] `cargo fmt` was run.
- [ ] Your changes were added to the `CHANGELOG.md` in the proper section.
- [x] You updated existing examples or added examples (if applicable).
- [ ] Added examples are checked in CI

### Nice to have

- [x] You add a description of your work to this PR.
- [ ] You added proper docs for your newly added features and code.

### CODE REVIEW

**Comment by JurajSadel** [comment]
@playfulFence IIRC you had a similar issue some long months ago, unfortunately, I forgot what your findings were.

**Comment by elpiel** [comment]
This leads me to believe that there might be something wrong with the board that I'm using.
Let me double check it tomorrow with another board.

**Comment by elpiel** [comment]
@bjoernQ I tried with another Rust board today and I'm still getting the same issue.

The only difference I see is the revision version of the chip, mine is 0.4 your is 0.3, not sure what difference does it make though.

```
[2023-11-11T09:01:13Z INFO ] Serial port: '/dev/ttyACM0'
[2023-11-11T09:01:13Z INFO ] Connecting...
[2023-11-11T09:01:13Z INFO ] Using flash stub
Chip type:         esp32c3 (revision v0.4)
Crystal frequency: 40MHz
Flash size:        4MB
Features:          WiFi, BLE
MAC address:       64:e8:33:dc:15:54
App/part. size:    294,000/4,128,768 bytes, 7.12%
[00:00:00] [========================================]      13/13      0x0                                                               [00:00:00] [========================================]       1/1       0x8000                                                            [00:00:02] [========================================]     117/117     0x10000                                                           [2023-11-11T09:01:16Z INFO ] Flashing has completed!
Commands:
    CTRL+R    Reset chip
    CTRL+C    Exit

ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x15 (USB_UART_CHIP_RESET),boot:0xc (SPI_FAST_FLASH_BOOT)
Saved PC:0x40380836
0x40380836 - esp_hal_common::interrupt::riscv::vectored::handle_interrupts
    at /mnt/Disk/Projects/lechev.space/esp-hal/esp-hal-common/src/interrupt/riscv.rs:216
SPIWP:0xee
mode:DIO, clock div:2
load:0x3fcd5820,len:0x171c
0x3fcd5820 - _stack_start
    at ??:??
load:0x403cc710,len:0x968
0x403cc710 - 
    at ??:??
load:0x403ce710,len:0x2f68
0x403ce710 - 
    at ??:??
SHA-256 comparison failed:
Calculated: 1d06b938c0222bf626e0bdf46178b1b37ab24d03f0360fc8fcf7153c2571deaf
Expected: 68d7bdf643ba446b8ed7ae8423241d442fd052b2bc77091100ba06fd65dcf8d5
Attempting to boot anyway...
entry 0x403cc710
0x403cc710 - 
    at ??:??
I (43) boot: ESP-IDF v5.1-beta1-378-gea5e0ff298-dirt 2nd stage bootloader
I (43) boot: compile time Jun  7 2023 07:59:10
I (44) boot: chip revision: v0.4
I (48) boot.esp32c3: SPI Speed      : 40MHz
I (53) boot.esp32c3: SPI Mode       : DIO
I (57) boot.esp32c3: SPI Flash Size : 4MB
I (62) boot: Enabling RNG early entropy source...
I (68) boot: Partition Table:
I (71) boot: ## Label            Usage          Type ST Offset   Length
I (78) boot:  0 nvs              WiFi data        01 02 00009000 00006000
I (86) boot:  1 phy_init         RF data          01 01 0000f000 00001000
I (93) boot:  2 factory          factory app      00 00 00010000 003f0000
I (101) boot: End of partition table
I (105) esp_image: segment 0: paddr=00010020 vaddr=3c040020 size=0a524h ( 42276) map
I (123) esp_image: segment 1: paddr=0001a54c vaddr=40380000 size=00e58h (  3672) load
I (124) esp_image: segment 2: paddr=0001b3ac vaddr=00000000 size=04c6ch ( 19564) 
I (134) esp_image: segment 3: paddr=00020020 vaddr=42000020 size=37c28h (228392) map
I (189) boot: Loaded app from partition at offset 0x10000
I (189) boot: Disabling RNG early entropy source...
 
 
!! A panic occured in 'examples/i2c_icm42670.rs', at line 45, column 76
 
PanicInfo {
    payload: Any { .. },
    message: Some(
        called `Result::unwrap()` on an `Err` value: BusError(ExecIncomplete),
    ),
    location: Location {
        file: "examples/i2c_icm42670.rs",
        line: 45,
        col: 76,
    },
    can_unwind: true,
}
 
Backtrace:
 
No backtrace available - make sure to force frame-pointers. (see https://crates.io/crates/esp-backtrace)
```

File: esp32c3-hal/examples/i2c_icm42670.rs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        io.pins.gpio8,
+        400u32.kHz(),
+        &clocks,
+    );
+
+    interrupt::enable(Interrupt::I2C_EXT0, interrupt::Priority::Priority1).unwrap();
+
+    use icm42670::accelerometer::Accelerometer;
+    let mut imu = icm42670::Icm42670::new(i2c, icm42670::Address::Primary).unwrap();
+
```
Comment (Ben-PH) [inline esp32c3-hal/examples/i2c_icm42670.rs:45]: Just a thought: Might be worth showing the setting of power levels, and a soft_reset followed by a 100ms delay.

I had to do some debugging when I did a software reset of the IMU: worked without the --release flag, but not when compiled in release. Adding a delay fixed it when I realized that was a datasheet thing.

**Comment by Ben-PH** [comment]
Can confirm it works on my machine. This is the binary that is on my device: 
[hello-imu.zip](https://github.com/esp-rs/esp-hal/files/13357580/hello-imu.zip)


(it's not actually a zip file, I had to rename it for github to accept it.)

That's compiled using `cargo run --release`. If you run `espflash flash --monitor /path/to/hello-imu` (after removing the `.zip`, it _should_ be a re-creation of what I've got running on my board (and working). If you still get the same issues, then probably a board thing. If no issues, I think it would be a toolchan, .toml, or something similar that's giving you troubles.

**Comment by elpiel** [comment]
It seems that the issue is more related the flashing and maybe timing.
I somehow managed to reset the device RST + BOOT and flash the firmware from @Ben-PH which worked a 2 times but then I couldn't replicate the fix.

Then flashing the firmware to the device after a BOOT, no matter if it's triggered with RST + BOOT or just holding BOOT when plugging in the USB gets me to this "waiting for download" and resetting the chip with `Ctrl + R` seems to get me to an I2C write reg mask, whatever this means...

```
    Finished release [optimized + debuginfo] target(s) in 0.44s
     Running `espflash flash --monitor target/riscv32imc-unknown-none-elf/release/examples/i2c_icm42670`
[2023-11-18T16:20:01Z INFO ] Serial port: '/dev/ttyACM0'
[2023-11-18T16:20:01Z INFO ] Connecting...
[2023-11-18T16:20:02Z INFO ] Using flash stub
Chip type:         esp32c3 (revision v0.4)
Crystal frequency: 40MHz
Flash size:        4MB
Features:          WiFi, BLE
MAC address:       64:e8:33:d8:c3:14
App/part. size:    116,656/4,128,768 bytes, 2.83%
[00:00:00] [========================================]      13/13      0x0                                                               [00:00:00] [========================================]       1/1       0x8000                                                            [00:00:01] [========================================]      40/40      0x10000                                                           [2023-11-18T16:20:04Z INFO ] Flashing has completed!
Commands:
    CTRL+R    Reset chip
    CTRL+C    Exit

ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x15 (USB_UART_CHIP_RESET),boot:0x4 (DOWNLOAD(USB/UART0/1))
Saved PC:0x40380832
0x40380832 - __EXTERNAL_INTERRUPTS
    at ??:??
waiting for download
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x15 (USB_UART_CHIP_RESET),boot:0x4 (DOWNLOAD(USB/UART0/1))
Saved PC:0x400462e0
0x400462e0 - rom_i2c_writeReg_Mask
    at ??:??
waiting for download
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x15 (USB_UART_CHIP_RESET),boot:0x4 (DOWNLOAD(USB/UART0/1))
Saved PC:0x400462dc
0x400462dc - rom_i2c_writeReg_Mask
    at ??:??
waiting for download
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x15 (USB_UART_CHIP_RESET),boot:0x4 (DOWNLOAD(USB/UART0/1))
Saved PC:0x400462dc
0x400462dc - rom_i2c_writeReg_Mask
    at ??:??
waiting for download
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x15 (USB_UART_CHIP_RESET),boot:0x4 (DOWNLOAD(USB/UART0/1))
Saved PC:0x400462e2
0x400462e2 - rom_i2c_writeReg_Mask
    at ??:??
waiting for download
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x15 (USB_UART_CHIP_RESET),boot:0x4 (DOWNLOAD(USB/UART0/1))
Saved PC:0x400462e2
0x400462e2 - rom_i2c_writeReg_Mask
    at ??:??
waiting for download
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x15 (USB_UART_CHIP_RESET),boot:0x4 (DOWNLOAD(USB/UART0/1))
Saved PC:0x400462e0
0x400462e0 - rom_i2c_writeReg_Mask
    at ??:??
waiting for download
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x15 (USB_UART_CHIP_RESET),boot:0x4 (DOWNLOAD(USB/UART0/1))
Saved PC:0x400462dc
0x400462dc - rom_i2c_writeReg_Mask
    at ??:??
waiting for download
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x15 (USB_UART_CHIP_RESET),boot:0x4 (DOWNLOAD(USB/UART0/1))
Saved PC:0x400462e0
0x400462e0 - rom_i2c_writeReg_Mask
    at ??:??
waiting for download
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x15 (USB_UART_CHIP_RESET),boot:0x4 (DOWNLOAD(USB/UART0/1))
Saved PC:0x400462e2
0x400462e2 - rom_i2c_writeReg_Mask
    at ??:??
waiting for download
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x15 (USB_UART_CHIP_RESET),boot:0x4 (DOWNLOAD(USB/UART0/1))
Saved PC:0x400462dc
0x400462dc - rom_i2c_writeReg_Mask
    at ??:??
waiting for download
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x15 (USB_UART_CHIP_RESET),boot:0x4 (DOWNLOAD(USB/UART0/1))
Saved PC:0x400462e0
0x400462e0 - rom_i2c_writeReg_Mask
    at ??:??
waiting for download
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x15 (USB_UART_CHIP_RESET),boot:0x4 (DOWNLOAD(USB/UART0/1))
Saved PC:0x400462e2
0x400462e2 - rom_i2c_writeReg_Mask
    at ??:??
waiting for download
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x15 (USB_UART_CHIP_RESET),boot:0x4 (DOWNLOAD(USB/UART0/1))
Saved PC:0x400462dc
0x400462dc - rom_i2c_writeReg_Mask
    at ??:??
waiting for download
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x15 (USB_UART_CHIP_RESET),boot:0x4 (DOWNLOAD(USB/UART0/1))
Saved PC:0x400462e2
0x400462e2 - rom_i2c_writeReg_Mask
    at ??:??
waiting for download
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x15 (USB_UART_CHIP_RESET),boot:0x4 (DOWNLOAD(USB/UART0/1))
Saved PC:0x400462e0
0x400462e0 - rom_i2c_writeReg_Mask
    at ??:??
waiting for download
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x15 (USB_UART_CHIP_RESET),boot:0x4 (DOWNLOAD(USB/UART0/1))
Saved PC:0x400462e0
0x400462e0 - rom_i2c_writeReg_Mask
    at ??:??
waiting for download
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x15 (USB_UART_CHIP_RESET),boot:0x4 (DOWNLOAD(USB/UART0/1))
Saved PC:0x400462dc
0x400462dc - rom_i2c_writeReg_Mask
    at ??:??
waiting for download
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x15 (USB_UART_CHIP_RESET),boot:0x4 (DOWNLOAD(USB/UART0/1))
Saved PC:0x400462e2
0x400462e2 - rom_i2c_writeReg_Mask
    at ??:??
waiting for download
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x15 (USB_UART_CHIP_RESET),boot:0x4 (DOWNLOAD(USB/UART0/1))
Saved PC:0x400462e2
0x400462e2 - rom_i2c_writeReg_Mask
    at ??:??
waiting for download
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x15 (USB_UART_CHIP_RESET),boot:0x4 (DOWNLOAD(USB/UART0/1))
Saved PC:0x400462e2
0x400462e2 - rom_i2c_writeReg_Mask
    at ??:??
waiting for download
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x15 (USB_UART_CHIP_RESET),boot:0x4 (DOWNLOAD(USB/UART0/1))
Saved PC:0x400462dc
0x400462dc - rom_i2c_writeReg_Mask
    at ??:??
waiting for download
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x15 (USB_UART_CHIP_RESET),boot:0x4 (DOWNLOAD(USB/UART0/1))
Saved PC:0x400462e2
0x400462e2 - rom_i2c_writeReg_Mask
    at ??:??
waiting for download
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x15 (USB_UART_CHIP_RESET),boot:0x4 (DOWNLOAD(USB/UART0/1))
Saved PC:0x400462e2
0x400462e2 - rom_i2c_writeReg_Mask
    at ??:??
waiting for download
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x15 (USB_UART_CHIP_RESET),boot:0x4 (DOWNLOAD(USB/UART0/1))
Saved PC:0x400462e2

```

**Comment by elpiel** [comment]
Same goes for @Ben-PH binary:
```
$ espflash flash --monitor hello-imu 
[2023-11-18T16:22:29Z INFO ] Serial port: '/dev/ttyACM0'
[2023-11-18T16:22:29Z INFO ] Connecting...
[2023-11-18T16:22:29Z INFO ] Using flash stub
Chip type:         esp32c3 (revision v0.4)
Crystal frequency: 40MHz
Flash size:        4MB
Features:          WiFi, BLE
MAC address:       64:e8:33:d8:c3:14
App/part. size:    111,504/4,128,768 bytes, 2.70%
[00:00:00] [========================================]      13/13      0x0                                                               [00:00:00] [========================================]       1/1       0x8000                                                            [00:00:01] [========================================]      38/38      0x10000                                                           [2023-11-18T16:22:31Z INFO ] Flashing has completed!
Commands:
    CTRL+R    Reset chip
    CTRL+C    Exit

ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x15 (USB_UART_CHIP_RESET),boot:0x4 (DOWNLOAD(USB/UART0/1))
Saved PC:0x40380836
0x40380836 - .LBB1_54
    at ??:??
waiting for download
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x15 (USB_UART_CHIP_RESET),boot:0x4 (DOWNLOAD(USB/UART0/1))
Saved PC:0x400462e2
0x400462e2 - rom_i2c_writeReg_Mask
    at ??:??
waiting for download

```

**Comment by georgik** [comment]
I was able to simulate the problem, by connecting ungrounded wire to PIN 8, or by touching the board.
It seems that I2C clocks are simply thrown out of balance.

Workaround: Decrease frequency of I2C to 2 kHz.

Probably it needs further investigation.

**Comment by jessebraham** [comment]
Thanks for reporting this problem, please see [this comment](https://github.com/esp-rs/esp-hal/issues/910#issuecomment-1827584906) for the cause of the problems.

I am going to close this PR for now, so we can carry on discussion in the linked issue.

**Comment by lu-zero** [comment]
I just tested the code and it seems running fine both in debug and release mode on the esp32c3 rust board.

**Comment by elpiel** [comment]
This was a hardware issue. The rust board has a problem because it uses two pins 8, which as far as I heard, causes the longer trace to act as an antenna.

If you put the board on a breadboards you will see the same results as me. On its own, the board will not show you this problem.

**Pull request closed without merging**